bazel/ci/def.bzl

```diff
@@ -0,0 +1,65 @@
+"""Bazel rules for CI and dev tooling"""
+
+load("@bazel_skylib//lib:shell.bzl", "shell")
+
+def _sh_template_impl(ctx):
+    out_file = ctx.actions.declare_file(ctx.label.name + ".bash")
+
+    substitutions = {}
+    for k, v in ctx.attr.substitutions.items():
+        sub = ctx.expand_location(v, ctx.attr.data)
+        sub = shell.quote(sub)
+        substitutions[k] = sub
+
+    ctx.actions.expand_template(
+        template = ctx.file.template,
+        output = out_file,
+        substitutions = substitutions,
+        is_executable = True,
+    )
+
+    return [DefaultInfo(
+        files = depset([out_file]),
+        executable = out_file,
+    )]
+
+_sh_template = rule(
+    implementation = _sh_template_impl,
+    attrs = {
+        "data": attr.label_list(
+            allow_files = True,
+        ),
+        "substitutions": attr.string_dict(),
+        "template": attr.label(
+            allow_single_file = True,
+        ),
+    },
+)
```

Comment:
This is defined in the global scope of this file, since it is not indented, right? Can we define this in a order which keeps those "bazel" definitions separate from the starlark functions?

Reply:
Just to clarify: I think the rule definition always has to live in the global scope of a `.bzl` file. [See also.](https://bazel.build/extending/rules)

The order can probably we changed though.